src/main/kotlin/org/rust/lang/core/types/infer/ProjectionCache.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.types.infer
+
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.ty.TyProjection
+
+/**
+ * The projection cache. Unlike the standard caches, this can
+ * include infcx-dependent type variables - therefore, we have to roll
+ * the cache back each time we roll a snapshot back, to avoid assumptions
+ * on yet-unresolved inference variables.
+ *
+ * Because of that, projection cache entries can be "stranded" and left
+ * inaccessible when type variables inside the key are resolved. We make no
+ * attempt to recover or remove "stranded" entries, but rather let them be
+ * (for the lifetime of the [RsInferenceContext]).
+ */
+class ProjectionCache {
+    private val map: SnapshotMap<TyProjection, ProjectionCacheEntry> = SnapshotMap()
+
+    fun startSnapshot(): Snapshot = map.startSnapshot()
+
+    /**
+     * Try to start normalize `key`. If we hits the cache (normalization
+     * already occurred), the function just returns the nonnull value from
+     * the cache. Otherwise it puts [ProjectionCacheEntry.InProgress] value
+     * to the cache and returns `null`
+     */
+    fun tryStart(key: TyProjection): ProjectionCacheEntry? {
+        return map.get(key) ?: run {
+            map.put(key, ProjectionCacheEntry.InProgress)
+            null
+        }
+    }
+
+    fun putTy(key: TyProjection, value: TyWithObligations<Ty>) {
+        map.put(key, ProjectionCacheEntry.NormalizedTy(value))
+            ?: error("never started projecting for `$key`")
+    }
+
+    fun ambiguous(key: TyProjection) {
+        map.put(key, ProjectionCacheEntry.Ambiguous)
+            ?: error("never started projecting for `$key`")
+    }
+
+    fun error(key: TyProjection) {
+        map.put(key, ProjectionCacheEntry.Error)
+            ?: error("never started projecting for `$key`")
```

Comment:
Maybe create something like
```kotlin
private fun put(key: TyProjection, value: ProjectionCacheEntry) {
    map.put(key, value) ?: error("never started projecting for `$key`")
}
```
and use it in `putTy`, `ambiguous` and `error` to avoid code duplication?